### PR TITLE
retrieving deployment error details during upgrade

### PR DESCRIPTION
**Pull request by dmitsh**
- retrieving deployment error details during upgrade

### CODE REVIEW

File: cmd/upgrade.go [hunk 1]
```diff
@@ -195,7 +195,7 @@ func (uc *upgradeCmd) run(cmd *cobra.Command, args []string) error {
 		Client:      uc.client,
 		StepTimeout: uc.timeout,
 	}
-	log.Fatalf("%v", upgradeCluster.StepTimeout)
```
Comment (JackQuincy) [inline cmd/upgrade.go]: opps thanks for removing my debugging line.

File: pkg/api/orchestrators.go [hunk 2]
```diff
@@ -181,6 +181,9 @@ func kubernetesUpgrades(csOrch *OrchestratorProfile) ([]*OrchestratorProfile, er
 	case currentMajor == 1 && currentMinor == 7:
 		nextMajor = 1
 		nextMinor = 8
+	case currentMajor == 1 && currentMinor == 8:
+		nextMajor = 1
```
Comment (JackQuincy) [inline pkg/api/orchestrators.go]: shouldn't this be a seperate PR as this feels unrelated and should be easy to push through on it's own.

File: pkg/operations/kubernetesupgrade/upgradeagentnode.go [hunk 3]
```diff
@@ -87,11 +87,19 @@ func (kan *UpgradeAgentNode) CreateNode(poolName string, agentNo int) error {
...
-	if err != nil {
-		return err
+	if err == nil {
+		return nil
 	}
 
-	return nil
+	kan.logger.Errorf("Deployment %s failed with error %v", deploymentName, err)
+	// Get deployment error details
+	errResp, e := armhelpers.GetDeploymentError(kan.Client, kan.logger, kan.ResourceGroup, deploymentName)
```
Comment (JackQuincy) [inline pkg/operations/kubernetesupgrade/upgradeagentnode.go]: if I remember right there are two types of errors here. One is a problem during the deployment and the other is during validation. This assumes all are the former. To figure out which kind it is we should need to look at the response etc as well as the error.

File: pkg/api/orchestrators.go (hunk 2 above)
Comment (dmitsh) [reply pkg/api/orchestrators.go]: agreed

File: pkg/armhelpers/deploymentError.go [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		// DeploymentOperationProperties contains error from one specific resource type so the
+		// chances of multiple deployment errors being returned for a single resource type is slim
+		// (but possible) based on current error/QoS analysis. In those cases where multiple errors
+		// are returned ACS will pick the first error code for determining whether this is an internal
+		// or a client error. This can be reevaluated later based on practical experience.
+		// However, note that customer will be returned the entire contents of "StatusMessage" object
+		// (like before) so they have access to all the errors returned by ARM.
+		logger.Infof("Found ResourceDeploymentFailure error code - error response = '%v'", *errresp)
+		details := errresp.Body.Details
+		if len(details) > 0 {
```
Comment (weinong) [inline pkg/armhelpers/deploymentError.go]: why do we want to lose the rest of error?

File: pkg/armhelpers/deploymentError.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+
+func toArmErrors(logger *logrus.Entry, deploymentName string, operationsList resources.DeploymentOperationsListResult) ([]*ErrorResponse, error) {
+	ret := []*ErrorResponse{}
+
+	if operationsList.Value == nil {
+		return ret, nil
+	}
+
+	for _, operation := range *operationsList.Value {
+		if operation.Properties == nil || *operation.Properties.ProvisioningState != string(api.Failed) {
```
Comment (weinong) [inline pkg/armhelpers/deploymentError.go]: can `ProvisioningState` be nil?

File: pkg/armhelpers/deploymentError.go [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+		errList = append(errList, eList...)
+	}
+	switch len(errList) {
+	case 0:
+		return nil, fmt.Errorf("unable to extract deployment error for %s", deploymentName)
+	case 1:
+		return errList[0], nil
+	default:
+		// combine all errors
+		for _, e := range errList[1:] {
```
Comment (weinong) [inline pkg/armhelpers/deploymentError.go]: why cant we just return an array of `*ErrorResponse`? Why do we want to change it?

File: pkg/armhelpers/deploymentError.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+// ErrorResponse  defines Resource Provider API 2.0 Error Response Content structure
+type ErrorResponse struct {
+	Body Error `json:"error"`
+}
+
+// Error implements error interface to return error in json
+func (e *ErrorResponse) Error() string {
+	return e.Body.Error()
```
Comment (anhowe) [inline pkg/armhelpers/deploymentError.go]: is this recursive at all?  I remember in C# there were errors stuffed inside other errors, and turtles all the way down.

Comment (dmitsh) [reply pkg/armhelpers/deploymentError.go]: yes, Error.Error() is recursive.

File: pkg/operations/kubernetesupgrade/upgradeagentnode.go [hunk 8]
```diff
@@ -87,11 +88,23 @@ func (kan *UpgradeAgentNode) CreateNode(poolName string, agentNo int) error {
...
+	kan.logger.Errorf("Deployment %s failed with error %v", deploymentName, err)
+	// Get deployment error details
+	errRespList, e := armhelpers.GetDeploymentError(kan.Client, kan.logger, kan.ResourceGroup, deploymentName)
+	if e != nil || len(errRespList) == 0 {
+		kan.logger.Errorf("Failed to get error details for deployment %s: %v", deploymentName, e)
+		// return original deployment error
+		return err
+	}
+	errStrList := make([]string, len(errRespList))
+	for i, errResp := range errRespList {
```
Comment (weinong) [inline pkg/operations/kubernetesupgrade/upgradeagentnode.go]: I thought you are changing it to return an array. why changed your mind?

Comment (dmitsh) [reply pkg/operations/kubernetesupgrade/upgradeagentnode.go]: I returned an array, but eventually have to turn it into an error

File: pkg/operations/kubernetesupgrade/upgradeagentnode.go [hunk 9]
```diff
@@ -80,18 +80,26 @@ func (kan *UpgradeAgentNode) CreateNode(poolName string, agentNo int) error {
 	deploymentSuffix := random.Int31()
 	deploymentName := fmt.Sprintf("agent-%s-%d", time.Now().Format("06-01-02T15.04.05"), deploymentSuffix)
 
-	_, err := kan.Client.DeployTemplate(
+	depExt, err := kan.Client.DeployTemplate(
```
Comment (JackQuincy) [inline pkg/operations/kubernetesupgrade/upgradeagentnode.go]: there is another possibility here. you are assuming it will work or there will be a list of deployment errors.  If the deployment fails early on, due to something like a quota issue, it will return the err in the response and won't have started any deployment operations. In the RP I have an example in arminteractor where we called deploytemplate there on how to find that error and parse it out.

File: pkg/apierror/const.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+	NetworkingInternalOperationError      ErrorCode = "NetworkingInternalOperationError"
+	QuotaExceeded                         ErrorCode = "QuotaExceeded"
+	Unauthorized                          ErrorCode = "Unauthorized"
+	ResourcesOverConstrained              ErrorCode = "ResourcesOverConstrained"
+	ControlPlaneProvisioningInternalError ErrorCode = "ControlPlaneProvisioningInternalError"
+	ControlPlaneProvisioningSyncError     ErrorCode = "ControlPlaneProvisioningSyncError"
+	ControlPlaneInternalError             ErrorCode = "ControlPlaneInternalError"
+	ControlPlaneCloudProviderNotSet       ErrorCode = "ControlPlaneCloudProviderNotSet"
+
+	// From Microsoft.WindowsAzure.ContainerService.API.AcsErrorCode
```
Comment (JackQuincy) [inline pkg/apierror/const.go]: the errors from here down appear to be RP errors. I'm not sure any of them apply in the ACS-Engine scenario. We should have a const file in the RP with these values and make sure that the datamodel down here just stores a string and matches these that way we can set values not defined here.

**Comment by amanohar** [comment]
My feedback and concerns about this PR remains the same as what we discussed in person:
1. The error codes are _very_ specific to what the RP and do not belong in ACS Engine at all
2. The error categories are not even an ARM concept and are internal to the RP. So while there is nothing confidential about it, it breaks abstraction and does not belong in ACS Engine
3. This change significantly impacts agility (and development cycle of the RP) of the RP OR we will end up with error codes in 2 places because a dev will just add to the RP to save time. 
4. My recommendations to you was to return ARM’s deployment results to the upgrade cmd and to the RP and RP can do the parsing.
